Throw an exception when MV columns are present in the order-by expression list in selection order-by only queries

somandal opened the pull request:
Today MV columns as the first expression in a Selection Order By query fails during query execution. The error seen today is:

```
18:25:22.961 BaseCombineOperator - Caught exception while processing query: QueryContext{_tableName='testTable', _subquery=null, _selectExpressions=[mvFloatCol], _aliasList=[null], _filter=mvFloatCol < '5', _groupByExpressions=null, _havingFilter=null, _orderByExpressions=[mvFloatCol ASC], _limit=10, _offset=0, _queryOptions={}, _expressionOverrideHints={}, _explain=false}
java.lang.ClassCastException: class [F cannot be cast to class java.lang.Comparable ([F and java.lang.Comparable are in module java.base of loader 'bootstrap')
	at org.apache.pinot.core.operator.combine.MinMaxValueBasedSelectionOrderByCombineOperator.processSegments(MinMaxValueBasedSelectionOrderByCombineOperator.java:210) ~[classes/:?]
	at org.apache.pinot.core.operator.combine.BaseCombineOperator$1.runJob(BaseCombineOperator.java:101) ~[classes/:?]
	at org.apache.pinot.core.util.trace.TraceRunnable.run(TraceRunnable.java:40) ~[classes/:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
	at java.lang.Thread.run(Thread.java:834) [?:?]
```

This PR throws an `UnsupportedOperationException` during the server side planning phase when it identifies a MV column as the first expression in the selection order-by query to provide users with a better message about why the query fails.

Note that this PR only errors out the scenario when the MV column is the first expression in the order-by expression list rather than erroring out whenever any MV column is present in the order-by expression list. The reason for this is to maintain backward compatibility. Today if a mix of SV and MV columns are present in a selection order-by query, and if the first expression is a SV column then the query execution completes successfully. The position of the MV column has no affect on the actual ordering of results as the `SelectionOrderByOperator` only includes the SV columns in the comparator used to set the values in the `PriorityQueue` maintained by it. The query execution failure only occurs when the MV column is the first expression.

Selection order-by on MV columns don't make sense semantically which is why it is better to error out in a meaningful way as the error thrown on the query execution failure requires debugging to understand why it failed.

cc @siddharthteotia

Jackie-Jiang commented:
Good catch!
I'd suggest doing it slightly different which can handle more general cases:
1. In `MinMaxValueBasedSelectionOrderByCombineOperator.MinMaxValueContext`, check if `dataSourceMetadata.isSingleValue()` and put `null` if it is not single valued so that we don't get exception
2. In `SelectionOrderByOperator.getComparator()`, we have access to the actual metadata of the transformed result, and we may perform the check if we don't want MV to be ordered (currently it skips MV expressions, but IMO it is okay to throw exception to prevent unexpected behavior)

walterddr commented:
> Good catch! I'd suggest doing it slightly different which can handle more general cases:
> 
> 1. In `MinMaxValueBasedSelectionOrderByCombineOperator.MinMaxValueContext`, check if `dataSourceMetadata.isSingleValue()` and put `null` if it is not single valued so that we don't get exception
> 2. In `SelectionOrderByOperator.getComparator()`, we have access to the actual metadata of the transformed result, and we may perform the check if we don't want MV to be ordered (currently it skips MV expressions, but IMO it is okay to throw exception to prevent unexpected behavior)

+1 on both suggestions. 

from this claim:
> The position of the MV column has no affect on the actual ordering of results as the SelectionOrderByOperator only includes the SV columns in the comparator used to set the values in the PriorityQueue maintained by it. The query execution failure only occurs when the MV column is the first expression.

I think we should definitely not allow MV values present in the order-by list with explicit error, and we can do that in the parser level already right?

somandal commented:
> Good catch! I'd suggest doing it slightly different which can handle more general cases:
> 
> 1. In `MinMaxValueBasedSelectionOrderByCombineOperator.MinMaxValueContext`, check if `dataSourceMetadata.isSingleValue()` and put `null` if it is not single valued so that we don't get exception
> 2. In `SelectionOrderByOperator.getComparator()`, we have access to the actual metadata of the transformed result, and we may perform the check if we don't want MV to be ordered (currently it skips MV expressions, but IMO it is okay to throw exception to prevent unexpected behavior)

Thanks for the suggestions @Jackie-Jiang 
I didn't understand what you meant by (1) above. What should I set to `null`?

For (2), we were thinking about throwing there, but there are a few problems.
a) We want to keep the explicit exception throwing backward compatible. Only identifiers go through the `MinMaxValueBasedSelectionOrderByCombineOperator` and already fail during query execution. Transform type expressions won't go through `MinMaxValueBasedSelectionOrderByCombineOperator` and if we now throw if the column is not a singleValue column then if anyone has existing queries with transform they may fail.
b) Did you want us to catch multi-value identifiers across all order-by expressions? for the same reason as (a) we wanted to avoid that because people might be running queries today where their first order by expression is single value identifier and with this change their query might fail.
c) We also wanted to catch this issue before starting the query execution. That's why we thought planning phase might be a good idea.

Let me know your thoughts on the above

amrishlal commented on pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java:
```diff
@@ -250,13 +250,33 @@ public PlanNode makeSegmentPlanNode(IndexSegment indexSegment, QueryContext quer
         return new AggregationPlanNode(indexSegment, queryContext);
       }
     } else if (QueryContextUtils.isSelectionQuery(queryContext)) {
+      validateSelectionOrderByExpressions(indexSegment, queryContext);
```
Is this the earliest stage that this check can be done? We should try to do such semantic checks as soon as possible (during or right after parsing) so that the query can be failed as quickly as possible without consuming any extra resources. There are a bunch of checks in BrokerRequestHandler, so I am wondering if this check can be done there instead of in server side plan generation?

Jackie-Jiang commented:
@somandal The idea here is to throw exception when user tries to order-by any expression that is MV. The earliest time when we know the expression is SV or MV is in the constructor of `SelectionOrderByOperator`, so we can catch it before the query execution. Actually we shouldn't need to change the `MinMaxValueBasedSelectionOrderByCombineOperator` because it is constructed in the execution phase, which is after the creation of `SelectionOrderByOperator`.

siddharthteotia commented:
So there are broadly 2 categories of MV specific order by expressions for selection order by queries 

- SELECT ..... ORDER BY <first expression is MV identifier expression> , 0 or more expressions could be SV or MV
- SELECT ..... ORDER BY <first expression is MV transform expression>, 0 or more expressions could be SV or MV

I agree that changing min-max operator is too late and the problem can be solved / detected sooner. 

I also agree that purist way of solving this is to detect the presence of **any** MV expression in the ORDER BY expressions like in `SelectionOrderByOperator` constructor and throw exception. I think it is a clean solution.

However, currently the engine lets 2nd category of queries go through by ignoring the MV expression in the ORDER BY list. The 1st category fails later during execution with not a meaningful error in the min-max operator when trying to execute first expression.

Making changes to constructor of `SelectionOrderByOperator` will fail both categories (which is good ) but I have no idea if our production has any queries from 2nd category  so there is a potential of breaking prod queries (if any) not just for us but even for OSS customers / users. There is no way for me to find percentage of such queries from our prod logs

To continue running 2nd category queries (like today), we decided not to make the operator change and instead fail it during server planning by special casing for category 1 queries.

Any thoughts on the risk to do the clean/pure solution ? If there is such query in production, we may have to rollback or force the client app to change queries. cc @somandal @Jackie-Jiang  @walterddr

Jackie-Jiang commented:
@siddharthteotia I understand the concern of breaking some prod queries, but IMO it is better to return exception rather than return unexpected result. The failed queries won't serve the purpose anyway, and returning an exception with proper error message is the best way to catch bad queries. Also, I don't think people have such queries, or they will complain when we added the `MinMaxValueBasedSelectionOrderByCombineOperator`

siddharthteotia commented:
@Jackie-Jiang @somandal 

I think we are on the same page here w.r.t what is the end state here -- block any MV expressions in ORDER BY via throwing exception in `SelectionOrderByOperator` constructor. 

I am not comfortable with making a change that may fail existing queries. Shall we do this ?

- Merge this change along with another change to detect presence of queries belonging to 2nd category (in my comment above) via a metric.
- Follow up in a couple of weeks time after having changed user queries (if any) depending on metric we observe internally and make the change in SelectionOrderByOperator for the general scenario. 

I am also ok with not doing this change at all and just do the metric change since when the final solution is implemented, the current change will probably be deleted anyway.

Jackie-Jiang commented:
It is a good idea to first emit metrics, then after monitoring for some time, change it to throw exceptions. I'd suggest only changing the `SelectionOrderByOperator` because that is the correct place to handle this check.

somandal commented:
> It is a good idea to first emit metrics, then after monitoring for some time, change it to throw exceptions. I'd suggest only changing the `SelectionOrderByOperator` because that is the correct place to handle this check.

Hey @Jackie-Jiang, I've created a new PR: https://github.com/apache/pinot/pull/9117 to address the metrics change. Once we get that in and monitor the usage on our end for such queries, I'll get back to this fix and make the clean change in the `SelectionOrderByOperator` directly. Can you take a look at that metrics change when you get a chance? Thanks!

somandal commented:
> It is a good idea to first emit metrics, then after monitoring for some time, change it to throw exceptions. I'd suggest only changing the `SelectionOrderByOperator` because that is the correct place to handle this check.

Hey @siddharthteotia and @Jackie-Jiang I've cleaned up this code, removed the metric and added the exception as part of the `SelectionOrderByOperator`. Can you folks take a look?

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/operator/query/SelectionOrderByOperator.java:
```diff
@@ -127,14 +125,22 @@ private Comparator<Object[]> getComparator() {
         valueIndexList.add(i);
       } else {
-        _queryHasMVSelectionOrderBy = true;
+        mvOrderByExpressions.add(_orderByExpressions.get(i));
```
(minor) We may directly throw exception here to have slightly better performance. IMO no need to gather all MV order-by expressions before throwing

somandal replied:
done

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/operator/query/SelectionOrderByOperator.java:
```diff
@@ -127,14 +125,22 @@ private Comparator<Object[]> getComparator() {
 
+    if (mvOrderByExpressions.size() > 0) {
+      // MV columns should not be part of the selection order by only list
+      throw new UnsupportedOperationException(
```
This should be `BadQueryRequestException`, so that we can gracefully handle it if necessary.

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/operator/query/SelectionOrderByOperator.java:
```diff
@@ -127,14 +125,22 @@ private Comparator<Object[]> getComparator() {
+    if (mvOrderByExpressions.size() > 0) {
+      // MV columns should not be part of the selection order by only list
+      throw new UnsupportedOperationException(
+          String.format("MV columns %s should not be included in selection order-by only queries",
```
```suggestion
          String.format("MV expression: %s should not be included in the ORDER-BY clause",
```

Jackie-Jiang reviewed (approved):
LGTM with minor comments
